File: javascript/src/text/numerals.js
/**
 * Bengali numeral conversion.
 *
 * Bengali uses its own digit glyphs (০-৯). These helpers convert between
 * Bengali and ASCII digits in both directions, leaving non-digits untouched.
 */

const BENGALI_DIGITS = "০১২৩৪৫৬৭৮৯";
const ASCII_DIGITS = "0123456789";

const BN_TO_EN = {};
const EN_TO_BN = {};
for (let d = 0; d < 10; d++) {
  BN_TO_EN[BENGALI_DIGITS[d]] = ASCII_DIGITS[d];
  EN_TO_BN[ASCII_DIGITS[d]] = BENGALI_DIGITS[d];
}

/**
 * Convert Bengali digits (০-৯) to ASCII digits (0-9).
 * @param {string} text
 * @returns {string}
 */
export function toEnglishDigits(text) {
  return text.replace(/[০-৯]/g, (c) => BN_TO_EN[c]);
}

/**
 * Convert ASCII digits (0-9) to Bengali digits (০-৯).
 * @param {string} text
 * @returns {string}
 */
export function toBengaliDigits(text) {
  return text.replace(/[0-9]/g, (c) => EN_TO_BN[c]);
}

File: python/bangla_ai/text/numerals.py
"""Bengali numeral conversion.

Bengali uses its own digit glyphs (০-৯, U+09E6–U+09EF). These utilities
convert between Bengali and ASCII (Western Arabic) digits in both directions,
leaving all non-digit characters untouched.
"""

from __future__ import annotations

_BENGALI_DIGITS = "০১২৩৪৫৬৭৮৯"
_ASCII_DIGITS = "0123456789"

_BN_TO_EN = str.maketrans(_BENGALI_DIGITS, _ASCII_DIGITS)
_EN_TO_BN = str.maketrans(_ASCII_DIGITS, _BENGALI_DIGITS)


def to_english_digits(text: str) -> str:
    """Convert Bengali digits (০-৯) to ASCII digits (0-9).

    >>> to_english_digits("২০২৪ সালে")
    '2024 সালে'
    """
    return text.translate(_BN_TO_EN)


def to_bengali_digits(text: str) -> str:
    """Convert ASCII digits (0-9) to Bengali digits (০-৯).

    >>> to_bengali_digits("2024 সালে")
    '২০২৪ সালে'
    """
    return text.translate(_EN_TO_BN)
